rfpremier/ah-django
# #164857217 Create, Edit & View Profiles

### What does this PR do?
This PR introduces adds the

 - [x] create profile,
 - [x] view specific profile ,
 - [x] view all profiles 
 - [x] edit my own profile functionality. to the profile app.

### Tasks to be completed

 - [x] Ensure a user profile is created on signup,
 - [x] view a specific profile ,
 - [x] view all profiles 
 - [x] Search through all profiles via user email , username and bio.
 - [x] Order through the profiles via username etc.
 - [x] Ensure a user can edit their own profiles alone.
 - [x] Add bio and image profile properties to login and signup response
 
 ### How can this be manually tests?
 
 - Clone the repo and checkout to the `ft/164857217-profiles-crud` branch
 - Run `python manage.py makemigrations`
 - Run `python manage.py migrate`
 - Create a new account with the following details
 `POST http://127.0.0.1:8000/api/users/`
     ```js
     {
     "user": {
       	"email": "profilemania@gmail.com",
		"password": "profiles2019",
		"username": "profmania"
    }
      }
     ```
     You will see the response (A new user profile  is created immediately after signup)
     ```js
	  {
	    "user": {
	        "email": "profilemania@gmail.com",
	        "username": "profmania",
	        "token": "tokenhere",
	        "bio": "",
	        "image": ""
	    }
	}
   ```
- To view all profiles
`GET http://127.0.0.1:8000/api/profiles/list/`

You will get the following results
```js
[
    {
        "image": "",
        "bio": "",
        "id": 24,
        "username": "profmania",
        "firstname": "",
        "lastname": "",
        "fullname": " "
    }
]
```

- To see a specific profile
`GET http://127.0.0.1:8000/api/profiles/<usernamehere>/`
`GET http://127.0.0.1:8000/api/profiles/profmania/`

You will get the following response
```js
{  
"profile": 
	  {
		  "image":  "",
		  "bio":  "",
		  "id":  24,
		  "username":"profmania",
                  "firstname": "",
                  "lastname": "",
                   "fullname": " "
	  }
  }
```

- To edit your profile
`PUT http://127.0.0.1:8000/api/profiles/profmania/`
You can edit either the 
- bio
- image
- firstname
- lastname
- or edit all the fields at once.
Remember to add your token in the `Authorization header`

```js
{
	"bio": "New Profmania bio"
}
```

You will get the response.

```js
{
    "profile": {
        "image": "",
        "bio": "New Profmania bio",
        "id": 24,
         "firstname": "",
          "lastname": "",
           "fullname": "",
        "username": "profmania"
    }
}
```

- Search through profiles by username or bio or user email address
`GET http://127.0.0.1:8000/api/profiles/list/?search=username`
`GET http://127.0.0.1:8000/api/profiles/list/?search=profmania`

```js
[
{
     "image":  "",
     "bio":  "New profmania bio",
     "id":  24,
     "username":  "profmania",
      "firstname": "",
        "lastname": "",
        "fullname": " "
     }
]
```
- Sort through profiles
  
	 - [x] **GET**  /api/profiles/list/?ordering=user__username  | Ascending
	 - [x] **GET**  /api/profiles/list/?ordering=-user__username | Descending
	 - [x] **GET**  /api/profiles/list/?ordering=bio | Ascending
	 - [x] **GET**  /api/profiles/list/?ordering=-bio | Descending


### Any background context
<strike>
In this app A username can be a single string with no space in between or a string with spaces in between.
If we have a username like
``` John Doe``` then we can update his profile for him via
```
*PUT*  http://127.0.0.1:8000/api/profiles/john-doe/
```
</strike>



The username now does not take any whitespace and the only alphanumeric it takes is the underscore.
The getting of a specific profile by username is now accurate but the is also not case sensitive as a username can have a mix of both capital letters and small letters

 ### PT stories relevant to this PR?
 [#164857217](https://www.pivotaltracker.com/n/projects/2321565/stories/164857217)
